### PR TITLE
tableinfo caching

### PR DESCRIPTION
**Pull request by labkey-matthewb**
Turn on caching in UserSchema.getTable() by default.
Enable foriegn keys to cache tableinfo as well.

### CODE REVIEW

**Comment by labkey-matthewb** [comment]
https://teamcity.labkey.org/project.html?projectId=LabKey_Trunk&branch_LabKey_Trunk=tableinfo_caching

File: experiment/src/org/labkey/experiment/api/LineageForeignKey.java [hunk 1]
```diff
@@ -65,7 +68,44 @@ public StringExpression getURL(ColumnInfo parent)
...
+
+        LevelColumnType(String name, String expType)
+        {
+            this.columnName = name;
+            this.expType = expType;
+        }
+
+        List<? extends ExpObject> getItems(UserSchema s)
+        {
+            // hmm don't think enums can override a method.. so here's a switch
```
[the remark this answers is not in the document]
Comment (labkey-matthewb) [reply experiment/src/org/labkey/experiment/api/LineageForeignKey.java]: yeah... I can't remember now what I was trying to do when I wrote that..

**Comment by labkey-jeckels** [comment]
I've been testing these changes, in conjunction with other optimizations for some TNPRC scenarios. Apologies for complicating this merge, but the combination is a massive win for complex and highly nested queries. This is at least a 40x speedup for some EHR queries.

**Comment by labkey-matthewb** [comment]
commented out code in AbstractTableCustomizer.getTableInfo()

**Comment by labkey-matthewb** [comment]
Love this PR!

**Comment by labkey-matthewb** [comment]
It would be nice to add a test to QueryTestCase with a performance counter (TestContext.logPerfResult())

**Comment by labkey-jeckels** [comment]
As for the QueryDefinition.getTable(UserSchema) question, I hadn't looked at that. Ideally we'd get rid of that argument, but I'd also like to get rid of the setSchema() method, especially on the QueryDefinition interface. But that would involve adding yet another overloaded version getQueryDef() and getQueryDefs() in QueryService which seems crummy

File: api/src/org/labkey/api/data/ContainerFilter.java [hunk 2]
```diff
@@ -985,6 +1036,12 @@ public InternalNoContainerFilter(User user)
             super(user);
         }
 
+        @Override
+        public String getCacheKey(Container c)
+        {
+            return "EVERYTHING";
```
[the remark this answers is not in the document]
Comment (labkey-matthewb) [reply api/src/org/labkey/api/data/ContainerFilter.java]: The obvious fix for this leads to a warning about potential classloader deadlock

EVERYTHING = new InternalNoContainerFilter()

Which is probably why I didn't fix it last time this bugged me.

**Pull request merged**